File: malai-core/org.malai.ts.react/src/interaction/library/ScrollDataImpl.ts
/*
 * This file is part of Malai.
 * Copyright (c) 2009-2018 Arnaud BLOUIN Gwendal DIDOT
 * Malai is free software; you can redistribute it and/or modify it under
 * the terms of the GNU General Public License as published by the Free Software
 * Foundation; either version 2 of the License, or (at your option) any later version.
 * Malai is distributed without any warranty; without even the implied
 * warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License for more details.
 */

import {ScrollData} from "./ScrollData";
import {Optional} from "../../util/Optional";

export class ScrollDataImpl implements ScrollData {

    protected scrolledNode: EventTarget | undefined;

    protected px: number | undefined;

    protected py: number | undefined;

    protected increment: number | undefined;

    public constructor() {}

    public reinitData(): void {
        this.increment = undefined;
        this.px = undefined;
        this.py = undefined;
        this.scrolledNode = undefined;
    }

    public getIncrement(): number {
        return this.increment === undefined ? 0 : this.increment;
    }

    public getPx(): number {
        return this.px === undefined ? 0 : this.px;
    }

    public getPy(): number {
        return this.py === undefined ? 0 : this.py;
    }

    public getScrolledNode(): Optional<EventTarget> {
        return Optional.ofNullable(this.scrolledNode);
    }

    public setScrollData(event: UIEvent) {
        this.scrolledNode = event.target === null ? undefined : event.target;
        this.increment = this.getIncrement() + (event.view.scrollY === undefined
        || event.view.scrollY < this.getIncrement() ? 0 : event.view.scrollY
            - (this.py === undefined || event.view.scrollY  < this.getIncrement() ? 0 : this.py));
        this.px = event.view.scrollX;
        this.py = event.view.scrollY;
    }
}

File: src/main/java/io/github/interacto/jfx/interaction/library/ScrollDataImpl.java
/*
 * Interacto
 * Copyright (C) 2020 Arnaud Blouin
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */
package io.github.interacto.jfx.interaction.library;

import javafx.scene.input.ScrollEvent;

/**
 * Scrolling interaction data implementation with write access.
 */
public class ScrollDataImpl implements ScrollData {
	/** The scrolled node. */
	protected Object scrolledNode;

	/** The X-coordinate of the scroll position. */
	protected double px;

	/** The Y-coordinate of the scroll position. */
	protected double py;

	/** The total increment of the scrolling. */
	protected double increment;

	@Override
	public Object getScrolledNode() {
		return scrolledNode;
	}

	@Override
	public double getPx() {
		return px;
	}

	@Override
	public double getPy() {
		return py;
	}

	@Override
	public double getIncrement() {
		return increment;
	}

	public void setData(final ScrollEvent event) {
		increment += event.getDeltaY();
		px = event.getX();
		py = event.getY();
		scrolledNode = event.getSource();
	}

	@Override
	public void flush() {
		px = 0d;
		py = 0d;
		increment = 0d;
		scrolledNode = null;
	}
}
